runtime/js-compute-runtime/builtins/crypto-algorithm.cpp

```diff
@@ -1105,6 +1221,260 @@ JSObject *CryptoAlgorithmHMAC_Import::toObject(JSContext *cx) {
+  // 1. Let keyData be the key data to be imported.
+  MOZ_ASSERT(cx);
+  JS::RootedObject result(cx);
+
+  std::unique_ptr<CryptoKeyECComponents> key = nullptr;
+  switch (format) {
+    // 2.  If format is "jwk":
+    case CryptoKeyFormat::Jwk: {
+      // 2.1. If keyData is a JsonWebKey dictionary:
+        // Let jwk equal keyData.
+      // Otherwise:
+        // Throw a DataError.
+      auto jwk = std::get<JsonWebKey *>(keyData);
+      if (!jwk) {
+        JS_ReportErrorLatin1(cx, "Supplied format is not a JSONWebKey");
+        return nullptr;
+      }
+      // 2.2. If the "d" field is present and usages contains a value which is not "sign",
+      //  or, if the "d" field is not present and usages contains a value which is not
+      // "verify" then throw a SyntaxError.
+      if (
+        (jwk->d.has_value() && !usages.isEmpty() && !usages.canOnlySign()) ||
+        (!jwk->d.has_value() && !usages.isEmpty() && !usages.canOnlyVerify())
+        ) {
+          DOMException::raise(cx, "EC keys only support 'sign' and 'verify' operations", "SyntaxError");
+          return nullptr;
+      }
+
+      // 2.3. If the "kty" field of jwk is not "EC", then throw a DataError.
+      // NOTE: This has already been done in the parent function.
+
+      // 2.4. If usages is non-empty and the "use" field of jwk is present and is not "sig", then throw a DataError.
+
+      if (!usages.isEmpty() && jwk->use.has_value() && jwk->use.value() != "sig") {
+        DOMException::raise(cx, "Operation not permitted", "DataError");
+        return nullptr;
+      }
+      // 2.5. If the "key_ops" field of jwk is present, and is invalid according to the requirements of JSON Web Key, or it does not contain all of the specified usages values, then throw a DataError.
+      if (jwk->key_ops.size() > 0) {
+        auto ops = CryptoKeyUsages::from(jwk->key_ops);
+        if (!ops.isSuperSetOf(usages)) {
+          DOMException::raise(cx,
+                              "The JWK 'key_ops' member was inconsistent with that specified by the "
+                              "Web Crypto call. The JWK usage must be a superset of those requested", "DataError");
+          return nullptr;
+        }
+      }
+      // 2.6. If the "ext" field of jwk is present and has the value false and extractable is true, then throw a DataError.
+      if (jwk->ext && !jwk->ext.value() && extractable) {
+        DOMException::raise(cx, "Data provided to an operation does not meet requirements", "DataError");
+        return nullptr;
+      }
+      // 2.7. Let namedCurve be a string whose value is equal to the "crv" field of jwk.
+      auto namedCurve = toNamedCurve(jwk->crv.value());
+      // 2.8. If namedCurve is not equal to the namedCurve member of normalizedAlgorithm, throw a DataError.
+      if (!namedCurve.has_value() || namedCurve.value() != this->namedCurve) {
+        DOMException::raise(cx, "Oopsie", "DataError");
```

Comment:
Should this error message be updated? (Though I do like "Oopsie" :D )

Reply:
lol oh yes it really should

Reply:
```suggestion
        DOMException::raise(cx, "The JWK's \"crv\" member specifies a different curve than requested", "DataError");
```